### ISSUE TITLE
Compiler error when filtering a ClosedRange of custom type

### ISSUE THREAD

**Issue by renaud-diabeloop** [groue/GRDB.swift#1416]
Hello everyone,

We have defined the following `Timestamp` type in our app, that we are trying to use everywhere possible, instead of `Foundation.Date`:

```swift
public struct Timestamp {
    /// Number of milliseconds since Epoch
    public let milliseconds: Int64
}
```

I added the conformance to `DatabaseValueConvertible`, very much like [in the documentation](https://swiftpackageindex.com/groue/grdb.swift/v6.16.0/documentation/grdb/databasevalueconvertible)

```Swift
extension Timestamp: DatabaseValueConvertible {
    public var databaseValue: GRDB.DatabaseValue {
        self.milliseconds.databaseValue
    }

    public static func fromDatabaseValue(_ dbValue: GRDB.DatabaseValue) -> Timestamp? {
        guard let milliseconds = Int64.fromDatabaseValue(dbValue) else {
            return nil
        }
        return Timestamp(milliseconds: milliseconds)
    }
}
```

Let's consider a `MealRecord` which uses a `Timestamp`:

```swift
struct MealRecord: Codable, Identifiable {
    var id: Int64?
    var date: Timestamp
}

extension MealRecord: TableRecord {
    static let databaseTableName: String = "Meal"

    /// Table columns
    enum Columns {
        static let id = Column(CodingKeys.id)
        static let date = Column(CodingKeys.date)
    }
}
```

My problem can be exposed in the following situation:

```Swift
func readRecords(betweenDates dateRange: ClosedRange<Timestamp>) async throws -> [MealRecord] {
    try await databaseQueue.read { db in
        try MealRecord
            .order(MealRecord.Columns.date)
            .filter(dateRange.contains(MealRecord.Columns.date)) // <-- ERROR
            .fetchAll(db)
    }
}
```

I get this error: 

> Cannot convert value of type 'Column' to expected argument type 'Timestamp'

Let's decompose this faulty line so the error is more precise:

```Swift
func readRecords(betweenDates dateRange: ClosedRange<Timestamp>) async throws -> [MealRecord] {
    try await databaseQueue.read { db in
        let column = MealRecord.Columns.date
        let predicate = dateRange.contains(column) // <-- ERROR

        try MealRecord
            .order(MealRecord.Columns.date)
            .filter(predicate)
            .fetchAll(db)
    }
}
```

> Cannot convert value of type 'Column' to expected argument type 'Timestamp'

Now, I kind of understand the issue. The compiler tries to call `ClosedRange<Timestamp>.contains(_ : Timestamp)`. This function does exist since `Timestamp` conforms to `Comparable`.

But it's not the right `contains` function. It should call this one from `SQLOperators.swift` instead:

```Swift
extension ClosedRange where Bound: SQLExpressible {
    public func contains(_ element: SQLSpecificExpressible) -> SQLExpression {
        .between(
            expression: element.sqlExpression,
            lowerBound: lowerBound.sqlExpression,
            upperBound: upperBound.sqlExpression)
    }
}
```

There are two things that I don't understand:

- `Timestamp` conforms to `SQLExpressible` through `DatabaseValueConvertible`. Why can't the compiler find it?
- Why does it work for `Date` ?

I know I still have the solution to stick to `Date`s and configure the `DatabaseDateEncodingStrategy`, but at least, I would like to understand. If you could provide some help, I would be grateful.

Thanks in advance, 
Renaud

### Environment

**GRDB flavor(s):** GRDB, SQLCipher,
**GRDB version:** 5.23.0
**Installation method:** CocoaPods
**Xcode version:** 14.3.1
**Swift version:** 5.1 ?
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** 13.5

**Comment by groue** [comment]
Salut Renaud 😙

This kind of error can happen in files that miss an `import GRDB`. I hope it is as simple as that!

And if it is not, please wait until next Tuesday, because I don't have any computer in front of me right now...

Bonnes bises depuis Stockholm!

**Comment by renaud-diabeloop** [comment]
Salut Gwendal 😗 

Unfortunately, my file does `import GRDB`.

Enjoy your trip to 🇸🇪 Sweden. In the meantime, I'll update GRDB to its latest version to see if it helps.

**Comment by groue** [comment]
Hello @renaud-diabeloop,

Maybe double-check that the file that contains the call to `contains` (where `readRecords` is defined) does import GRDB.

I can't reproduce this compiler error. Here is a sample code below that compiles and runs fine (in a playground):

```swift
import GRDB

public struct Timestamp {
    /// Number of milliseconds since Epoch
    public let milliseconds: Int64
}

extension Timestamp: Comparable {
    public static func < (lhs: Timestamp, rhs: Timestamp) -> Bool {
        lhs.milliseconds < rhs.milliseconds
    }
}

extension Timestamp: Codable {
    public init(from decoder: Decoder) throws {
        self.milliseconds = try decoder.singleValueContainer().decode(Int64.self)
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(milliseconds)
    }
}

extension Timestamp: DatabaseValueConvertible { /* inferred from Codable */ }

struct MealRecord: Identifiable, Codable {
    var id: Int64?
    var date: Timestamp
}

extension MealRecord: FetchableRecord, MutablePersistableRecord {
    static let databaseTableName: String = "meal"
    
    enum Columns {
        static let id = Column(CodingKeys.id)
        static let date = Column(CodingKeys.date)
    }
    
    mutating func didInsert(_ inserted: InsertionSuccess) {
        id = inserted.rowID
    }
}

// ----

var configuration = Configuration()
configuration.publicStatementArguments = true
configuration.prepareDatabase { db in
    db.trace { print("SQL> \($0)") }
}
let dbQueue = try DatabaseQueue(configuration: configuration)

func readRecords(betweenDates dateRange: ClosedRange<Timestamp>) async throws -> [MealRecord] {
    try await dbQueue.read { db in
        let column = MealRecord.Columns.date
        let predicate = dateRange.contains(column) // <-- NO ERROR

        return try MealRecord
            .order(MealRecord.Columns.date)
            .filter(predicate)
            .fetchAll(db)
    }
}


try dbQueue.write { db in
    try db.create(table: "meal") { t in
        t.autoIncrementedPrimaryKey("id")
        t.column("date", .integer).notNull()
    }
    
    do {
        var meal = MealRecord(id: nil, date: Timestamp(milliseconds: 123))
        try meal.insert(db)
    }
    
    do {
        let dateRange: ClosedRange<Timestamp> = Timestamp(milliseconds: 0)...Timestamp(milliseconds: 1000)
        let meals = try MealRecord
            .filter(dateRange.contains(MealRecord.Columns.date)) // <-- NO ERROR
            .fetchAll(db)
        for meal in meals {
            print(meal)
        }
    }
}
```

**Comment by renaud-diabeloop** [comment]
Hello @groue. I checked twice and thrice, but GRDB really is imported.
I have upgraded GRDB and cleaned Derived data (you never know).

I know get two errors for the same `filter()` line:
> Cannot convert value of type 'Column' to expected argument type 'Timestamp'
> Missing argument label 'key:' in call

I'm gonna try you example and let you know.

Thanks you indeed for taking the time to investigate the issue!

**Comment by renaud-diabeloop** [comment]
Okay, I found the problem out!

Something I had omitted for the sake of simplicity is that `Timestamp` is declared in an other module (SPM package); let's call it `DBLTime`.

Since I don't want the `DBLTime` module to depend on GRDB, conformance of `Timestamp` to `DatabaseValueConvertible` is done in the "App" module:

```Swift
extension DBLTime.Timestamp: DatabaseValueConvertible {
    public var databaseValue: DatabaseValue {
        self.milliseconds.databaseValue
    }

    public static func fromDatabaseValue(_ dbValue: DatabaseValue) -> Timestamp? {
        guard let milliseconds = Int64.fromDatabaseValue(dbValue) else {
            return nil // The database value is not an Int64
        }
        return .init(milliseconds: milliseconds)
    }
}
```

My mistake was that this file was not added to the "App" target 🤦. You know how it happens: you want to start from a clean state, so you extract the file from the projet, do a `git reset --hard`, put back the file in place… and forget to add it to the target.
It's not the first time that it happens to me, but understanding was extra difficult. Usually you find out quickly as the symbol is just not known to the compiler, but here it was.

In short, it's a stupid mistake — as always — I don't know if there is a lesson here, but make sure you checked the target 😓 

Thanks again for the help!
Renaud

**Comment by groue** [comment]
Great, well done Renaud 🙂

This recent thread on the Swift forums also talks about the time spent understanding the cause behind a poor compiler diagnostic: https://forums.swift.org/t/error-when-attempting-to-use-non-public-library-member-is-too-opaque-and-could-be-more-helpful/66840
